src/rust/engine/fs/src/pool.rs

```diff
@@ -40,31 +46,24 @@ impl ResettablePool {
     R::Item: Send + 'static,
     R::Error: Send + 'static,
   {
-    {
-      // The happy path: pool is already initialized.
-      let pool_opt = self.pool.read().unwrap();
-      if let Some(ref pool) = *pool_opt {
-        return pool.spawn_fn(f);
-      }
-    }
-    {
-      // Initialize the pool, but then release the write lock.
-      let mut pool_opt = self.pool.write().unwrap();
-      pool_opt.get_or_insert_with(|| self.new_pool());
-    }
-
-    // Recurse to run the function under the read lock.
-    self.spawn_fn(f)
+    let pool_opt = self.pool.read().unwrap();
+    let pool = pool_opt.as_ref().unwrap_or_else(|| {
+      panic!("A CpuPool cannot be used inside the fork context.")
+    });
+    return pool.spawn_fn(f);
   }
 
-  pub fn reset(&self) {
+  ///
+  /// Run a function while the pool is shut down, and restore the pool after it completes.
+  ///
+  pub fn with_shutdown<F, T>(&self, f: F) -> T
```

Comment:
Could we do away with the lock entirely by making `with_shutdown` take `&mut self`?

(I can believe this is impractical, but it would be nice if possible...)

Reply:
I don't think so, no... we have a reference to the pool via an `Arc`, and getting a mutable reference into that would require either cloning or something potentially panicy.